### PR TITLE
Merge MuSig2 functions into ekrem/new-architecture

### PR DESCRIPTION
**Pull request by ozankaymak**
# Description
This PR introduces the complete musig2 functionality, built on top of this
[musig2 implementation](https://docs.rs/musig2/latest/musig2/).

## Linked Issues
Related to #205.

## Testing
Unit tests, dummy transaction verification tests, and rpc tests are implemented.

## Docs
Code is documented.

### CODE REVIEW

File: core/Cargo.toml [hunk 1]
```diff
@@ -28,6 +28,7 @@ clap = { workspace = true, features = ["derive"] }
 toml = { workspace = true }
 sqlx = { workspace = true, features = ["runtime-tokio", "postgres"] }
 bitcoin-mock-rpc = { workspace = true }
+musig2 = { version = "0.0.11" }
```
Comment (ekrembal) [inline core/Cargo.toml:31]: we can use workspace=true here

File: core/src/errors.rs [hunk 2]
```diff
@@ -224,3 +227,9 @@ impl From<sqlx::Error> for BridgeError {
         BridgeError::DatabaseError(err)
     }
 }
+
+impl From<musig2::errors::KeyAggError> for BridgeError {
+    fn from(_err: musig2::errors::KeyAggError) -> Self {
+        BridgeError::KeyAggContextError
```
Comment (ekrembal) [inline core/src/errors.rs:233]: Pls propogate the original message in this

**Review by ekrembal — COMMENTED** [review]
I think whats left is aggregate nonces and partial signatures. And tests also

**Comment by ekrembal** [comment]
We also need tweaked signatures as well

File: core/src/musig2.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+use crate::{errors::BridgeError, ByteArray66};
+
+// We can directly use the musig2 crate for this
+// No need for extra types etc.
+// MuSigPubNonce consists of two curve points, so it's 66 bytes (compressed).
+pub type MuSigPubNonce = ByteArray66;
+// MuSigSecNonce consists of two scalars, so it's 64 bytes.
+pub type MuSigSecNonce = [u8; 64];
+// MuSigAggNonce is a scalar, so it's 32 bytes.
+pub type MuSigAggNonce = [u8; 32];
```
Comment (ekrembal) [inline core/src/musig2.rs:13]: This should be two curve points

File: core/src/musig2.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    .unwrap()
+}
+
+// Giving Vec<PublicKey> as an argument since we need to find the index of the keypair in the list of public keys;
+// otherwise, we can simply pass it as an argument. Also we need more entropy for the nonce.
+pub fn nonce_pair(
+    keypair: &secp256k1::Keypair,
+    rng: &mut impl Rng,
+    pks: Vec<PublicKey>,
+    tweak: Option<[u8; 32]>,
```
Comment (ekrembal) [inline core/src/musig2.rs:48]: we don't need tweak and pks here

**Comment by ekrembal** [comment]
I think we need 3 tests:
- one for signing a 32 byte hash and verifying the agg signature with agg pubkey
- one for signing a taproot key spend that does have some script spending paths so tweak is needed
- one for signing a taproot from a script spending path with aggPubkey checkSig

File: core/src/musig2.rs [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+        &key_agg_ctx,
+        musig2::secp256k1::SecretKey::from_slice(&keypair.secret_key().secret_bytes()).unwrap(),
+        musig_sec_nonce,
+        &musig_agg_nonce,
+        &sighash,
+    )
+    .unwrap();
+    partial_signature
+}
+
```
Comment (ekrembal) [inline core/src/musig2.rs:91]: Let's add aggregate_signatures and aggregate_nonces here as well

File: core/src/transaction_builder.rs [hunk 6]
```diff
@@ -282,6 +287,43 @@ impl TransactionBuilder {
...
+        pks: Vec<PublicKey>,
+        scripts: Vec<ScriptBuf>,
+        network: bitcoin::Network,
+    ) -> Result<(Address, TaprootSpendInfo), BridgeError> {
+        let n = scripts.len();
+        if n == 0 {
+            return Err(BridgeError::TaprootScriptError);
+        }
+
+        let taproot_builder = if n > 1 {
```
Comment (ekrembal) [inline core/src/transaction_builder.rs:301]: There is a code duplication here

**Review by ekrembal — CHANGES_REQUESTED** [review]

File: core/src/transaction_builder.rs (hunk 6 above)
Comment (ozankaymak) [reply core/src/transaction_builder.rs:301]: I do not understand what you mean.

**Review by ekrembal — APPROVED** [review]

**Pull request merged**